### ISSUE TITLE
Get full probability distribution out of yolov8

### ISSUE THREAD

**Issue by EricPedley** [uci-uav-forge/uavf_2024#12]
# Situation
During the mission, we want to be able to choose a target to drop every payload on, even if the description we detect doesn't match our payload on all 4 criteria (shape, symbol, shape color, symbol color). To do this, we would like if instead of just detecting the 4 labels for every target, we got probability distributions across the label classes (e.g. 60% circle, 20% rectangle, 10% triangle). 

Right now our shape detector and letter classifier only give us the highest confidence classification and its probability, but not the rest. However, it does calculate these at some point. If you do `model.model(data)` instead of `model.predict(data)`, we can get the raw output of the model, but at the cost of it not being processed completely. For example, the shape detector outputs a `(1,49,8400)` tensor using this method. The first dimension is the batch size and the last one is the number of total boxes predicted. The middle dimension is split up like this:
`x,y,w,h, (13-class distribution), (32 mask weights that we can ignore)`
"mask weights" is kinda explained here if anyone is curious (thx Eesh for finding this): https://github.com/ultralytics/ultralytics/issues/2953

# Task
Figure out how to get the full class distributions instead of only the highest confidence. This probably involves using our own NMS to filter boxes, or reuse whatever we develop in #11. The current approach is here: https://github.com/uci-uav-forge/uavf_2024/blob/01db54e2dfec82bdb5cb8620041793ee7951139f/uavf_2024/imaging/shape_detection/shape_instance_segmenter.py#L34C1-L35C42

Verify that this works by pausing the debugger on a breakpoint somewhere in that function and taking a screenshot of the full class probability, then put that in the PR you open.

**Comment by MinhxNguyen7** [comment]
Not exactly sure what you mean by `model.model(data)` because that seems to just be the path to the model.
<img width="521" alt="image" src="https://github.com/uci-uav-forge/uavf_2024/assets/64875104/7f311303-d4e7-47e0-a474-ee631cee3beb">

**Comment by MinhxNguyen7** [comment]
I've been running `ShapeInstanceSegmenter` in in the debugger and looking around. There are no type hints anywhere, and nothing makes sense. Here's my notes dump.

`shape_model.model` is class `SegmentationModel`, which is callable because it's a descendent of Torch's Module (`torch.nn.modules.module.Module`). However, calling `shape_model.model(img)` instead of `shape_model.predict(img)` results in a type error. Maybe I just need to convert it to a Tensor.
<img width="603" alt="image" src="https://github.com/uci-uav-forge/uavf_2024/assets/64875104/8cc5c9bc-f1d0-454b-b82c-7ccb733387de">

It needed to be transposed from `(height, width, channels)` or `(width, height, channels)` to `(channels, height, width)` or `(channels, width, height)`. I haven't looked through the code enough to be able to tell, especially because it says different things in different places. In any case, I can confirm that I get a tensor of shape (1,49,8400).
![image](https://github.com/uci-uav-forge/uavf_2024/assets/64875104/02a0d604-c236-4fa2-a316-1030767a3eb1)

I'll try to extract the class distribution next, which I understand is the part that you actually needed doing.
